src/Datadog.Trace/Agent/MovingAverageKeepRateCalculator.cs

```diff
@@ -84,14 +86,8 @@ public void IncrementDrops(int count)
         /// </summary>
         public double GetKeepRate()
         {
-            var totals = Interlocked.Read(ref _totals);
-            UnpackBits(totals, out var totalDropped, out var totalCreated);
-            if (totalCreated == 0)
-            {
-                return 0;
-            }
-
-            return 1 - ((double)totalDropped / totalCreated);
+            // Essentially Interlock.Read(double)
+            return Interlocked.CompareExchange(ref _keepRate, 0, 0);
```

Comment:
```suggestion
            return Volatile.Read(ref _keepRate);
```

This will emit the same cmpxchg on x86, but more optimized code on x64

Reply:
Really?! Everything I read suggested that `double` read/write wasn't atomic with Volatile 🤔 e.g. [StackOverflow is lying to me](https://stackoverflow.com/questions/12435325/does-volatile-read-volatile-write-for-double-atomic)?

Reply:
Stackoverflow is talking about the `volatile` keyword, which indeed doesn't guarantee atomicity. The `Volatile` class is different. It's implemented as an Interlocked operation and replaced by the JIT on architectures that allow it: https://referencesource.microsoft.com/#mscorlib/system/threading/volatile.cs,189
Atomicity is guaranteed.

Reply:
`Volatile.Read` is implemented as an atomic operation (on properly aligned data, so all bets are off if you start using `StructLayout`). It uses Interlocked on 32 bit platforms, and native read/write on 64 bit platforms: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Threading/Volatile.cs#L91

That said, it's true that this atomicity is not documented on MSDN.

Reply:
You're right of course - I hadn't twigged they were talking about the volatile keyword rather than the class. I can hardly find anything about the class!

Don't suppose you have a blog post about it do you? 😉

Reply:
Do we need to worry about atomicity? i.e. is this code ever called from multiple threads concurrently?

Reply:
> Do we need to worry about atomicity? i.e. is the read/write code ever called from multiple threads concurrently?

Yes 🙂 the update loop runs on a separate thread to the agent which is returning the value